include/zephyr/bluetooth/gatt.h

```diff
@@ -1257,15 +1257,30 @@ struct bt_gatt_discover_params;
  *  Only the following fields of the attribute contains valid information:
  *   - uuid      UUID representing the type of attribute.
  *   - handle    Handle in the remote database.
- *   - user_data The value of the attribute.
- *               Will be NULL when discovering descriptors
- *
- *  To be able to read the value of the discovered attribute the user_data
- *  must be cast to an appropriate type.
- *   - @ref bt_gatt_service_val when UUID is @ref BT_UUID_GATT_PRIMARY or
- *     @ref BT_UUID_GATT_SECONDARY.
- *   - @ref bt_gatt_include when UUID is @ref BT_UUID_GATT_INCLUDE.
- *   - @ref bt_gatt_chrc when UUID is @ref BT_UUID_GATT_CHRC.
+ *   - user_data The value of the attribute, if the discovery type maps to an
+ *               ATT operation that provides this information. NULL otherwise.
+ *               See below.
+ *
+ *  The following is a list of the possible combinations of @c params->type as
+ *  given to @ref bt_gatt_discover, the @c params->uuid argument to this
+ *  callback and the resulting type of @c user_data.
```

Comment:
Thanks :)

https://github.com/zephyrproject-rtos/zephyr/commit/50861254140c064bd6a049cc15985c7dabf64829 is incorrect. `bt_gatt_attr_func_t` is for iterating over local attributes. For local attributes, the `user_data` field is actually a user-data field, which the Host does not access. The documentation for `bt_gatt_foreach_attr_type` parameter `attr_data` can however be improved. That parameter is checked for equality with `attr.user_data` when filtering, except when it's `NULL`. This is not clear from the documentation. I believe this should go in a separate PR.

Reply:
You are right. However it is more correct to say that `bt_gatt_attr.read == bt_gatt_attr_read_chrc` determines the type of `user_data` to be `bt_gatt_chrc`. It's possible for some app/module to define a `BT_GATT_PRIMARY_SERVICE` that is not managed by the host. Then the `user_data` may have a different meaning. And anyway, shouldn't how `user_data` is used be an implementation detail of `BT_GATT_PRIMARY_SERVICE`, etc? The value be read trough `bt_gatt_attr.read` after all, (and then the format is determined by the type of the attribute).

Reply:
Indeed. 

If a GATT attribute is defined without the macros, then the type assumption no longer holds, so I agree that the `read` or `write` functions are better at determining the type. However if the GATT attributes are not created with the macros, all hell breaks lose as the stack (gatt.c) will in many places expect that `attr->user_data` to be specific types.

Reply:
If all hell breaks loose, can we document that somewhere ? That not using the macros is not supported and may break assumptions in the stack.

Reply:
At least in `gatt.c` we have the following lines without any confirmation:
```c
struct bt_uuid *uuid = attr->user_data;
struct bt_gatt_attr *incl = attr->user_data;
struct bt_gatt_chrc *chrc = attr->user_data;
struct _bt_gatt_ccc *ccc = attr->user_data;
const struct bt_gatt_cep *value = attr->user_data;
const struct bt_gatt_cpf *cpf = attr->user_data;
```

Basically the only confirmation we have is 
```c
if (attr->write != bt_gatt_attr_write_ccc) {
```
for the CCCDs, and even that is not applied everywhere.

Reply:
> ```c
> const struct bt_gatt_cpf *cpf = attr->user_data;
> ```

You are right, we do actually document the type in those, e.g. by ` *  @note Only use this with attributes which user_data is a bt_gatt_include.`, and we can argue that since it is documented, the erroneous behavior of an application doing 
```c
	BT_GATT_ATTRIBUTE(BT_UUID_GATT_INCLUDE, BT_GATT_PERM_READ,	\
			  bt_gatt_attr_read_included, NULL, NULL)
```
(i.e. supplying NULL instead of the service_include) is against the documented requirement, but it will still build and run until some client reads the included services, causing a fatal crash that may be very hard to debug for an application. 


In any case, the idea that the foreach will return `attr->user_data` of a specific type is not true, as you pointed out @alwa-nordic, as an application can, for whatever reason, actually do 
```c
	BT_GATT_ATTRIBUTE(BT_UUID_GATT_INCLUDE, BT_GATT_PERM_READ,	\
			  app_read_included, NULL, NULL)
```
where the `attr->user_data` is not a service UUID, and may actually still work. The `attr->user_data` assumption is only true for attributes defined by the macros. 

@alwa-nordic Do you suggest we close the other PR and leave it as is, or can we somehow add enough documentation for the `attr->user_data` usable?

Reply:
> @alwa-nordic Do you suggest we close the other PR and leave it as is, or can we somehow add enough documentation for the attr->user_data usable?

Documenting the type of `attr->user_data` is adding a new contract that was not there when it was unspecified. I'm not wholly against this, but we have to think carefully about doing this because 1. it may break existing applications (e.g. when a 3. party has defined a characteristic), and 2. it arguably exposes an internal detail of the host (We need to convince ourselves that we will never want to change the format of user_data).

I prefer if we don't add this documentation in this PR at least.